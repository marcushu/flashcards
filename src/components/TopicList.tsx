interface TopicListProps {
  usersTopics: string[] | []
  setTopic: (newTopic: string) => void
  deleteTopic: (_topic: string) => void
  currentTopic: string
  userStatus: string
}

const TopicList = ({ usersTopics, setTopic, deleteTopic, currentTopic, userStatus }: TopicListProps) => {
  return (
    <>
      {!!usersTopics.length &&
        <div>
          {(usersTopics as string[]).map((topic: string) =>
            <p key={topic}>

              <button 
                onClick={() => setTopic(topic)}
                className={topic === currentTopic ? "selected" : ""}>
                {topic}
              </button>
              {topic === currentTopic && userStatus !== 'GUEST' &&
                <button 
                  onClick={() => deleteTopic(topic)}
                  style={{ color: "red", fontSize: "12px" }}>
                  delete
                </button>}
            </p>
          )}
        </div>
      }
    </>
  )
}

export default TopicList