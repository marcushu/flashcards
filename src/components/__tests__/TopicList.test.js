import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import TopicList from '../TopicList';

describe("TopicList", () => {
    const mockDeleteFn = jest.fn();
    const mockSetFn = jest.fn();

    it("renders", () => {
        const { asFragment } = render(
            <TopicList currentTopic={"java"}
                usersTopics={["java", "c", "fortran"]}
                deleteTopic={mockDeleteFn}
                setTopic={mockSetFn}
                userStatus={"user"} />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("displays the list of topics", () => {
        const { getByRole } = render(
            <TopicList currentTopic={"java"}
                usersTopics={["java", "c", "fortran"]}
                deleteTopic={mockDeleteFn}
                setTopic={mockSetFn}
                userStatus={"user"} />
        );

        expect(getByRole('button', {  name: /java/i})).toBeInTheDocument();
        expect(getByRole('button', {  name: /c/i})).toBeInTheDocument();
        expect(getByRole('button', {  name: /fortran/i})).toBeInTheDocument();
    });

    it("displays the correct element as selected", () => {
        const { getByRole } = render(
            <TopicList currentTopic={"java"}
                usersTopics={["java", "c", "fortran"]}
                deleteTopic={mockDeleteFn}
                setTopic={mockSetFn}
                userStatus={"user"} />
        );

        expect(getByRole('button', {  name: /java/i})).toHaveClass("selected");
    });

    it("calls delete and set functions with the correct argument", () => {
        const { getByRole } = render(
            <TopicList currentTopic={"java"}
                usersTopics={["java", "c", "fortran"]}
                deleteTopic={mockDeleteFn}
                setTopic={mockSetFn}
                userStatus={"user"} />
        );
        
        fireEvent.click(getByRole('button', {  name: /delete/i}));
        expect(mockDeleteFn).toHaveBeenCalledWith("java");

        fireEvent.click(getByRole('button', {  name: /c/i}));
        expect(mockSetFn).toHaveBeenCalledWith("c");
    })
});