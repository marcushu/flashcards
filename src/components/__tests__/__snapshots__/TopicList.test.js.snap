// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicList renders 1`] = `
<DocumentFragment>
  <div>
    <p>
      <button
        class="selected"
      >
        java
      </button>
      <button
        style="color: red; font-size: 12px;"
      >
        delete
      </button>
    </p>
    <p>
      <button
        class=""
      >
        c
      </button>
    </p>
    <p>
      <button
        class=""
      >
        fortran
      </button>
    </p>
  </div>
</DocumentFragment>
`;
